File: src/main/java/doom/despair/core/GameEvent.java
package doom.despair.core;
import com.google.gson.Gson;

public abstract sealed class GameEvent permits ClientGameEvent, ServerGameEvent {
    public String message;
    public String type;
    public final String serialize() {
        return new Gson().toJson(this);
    }
    public static <T extends GameEvent> T deserialize(String json, Class<T> clazz) {
        return new Gson().fromJson(json, clazz);
    }
    public GameEvent() {
        this.type = this.getClass().getSimpleName();
    }
}

File: src/main/java/doom/despair/core/GameEvent.kt
package doom.despair.core

import com.google.gson.*;

abstract class GameEvent {
    @JvmField
    var message: String? = null
    @JvmField
    var type: String = this.javaClass.getSimpleName()
    fun serialize(): String {
        return Gson().toJson(this)
    }

    companion object {
        fun <T : GameEvent?> deserialize(json: String?, clazz: Class<T>): T {
            return Gson().fromJson(json, clazz)
        }
    }
}
